File: app/src/main/java/com/mintthursday/recipe/creation/NewRecipeFragment.java
package com.mintthursday.recipe.creation;

import static com.mintthursday.recipe.creation.ingredientcreation.IngredientFragment.BUN_INGREDIENT;
import static com.mintthursday.recipe.creation.ingredientcreation.IngredientFragment.BUN_ITEM_POSITION;
import static com.mintthursday.recipe.creation.ingredientcreation.IngredientFragment.REQ_INGREDIENT;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentResultListener;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.mintthursday.App;
import com.mintthursday.MainActivity;
import com.mintthursday.R;
import com.mintthursday.database.AppDatabase;
import com.mintthursday.database.RecipeDao;
import com.mintthursday.models.Ingredient;
import com.mintthursday.models.Recipe;
import com.mintthursday.models.Step;
import com.mintthursday.recipe.creation.ingredientcreation.IngredientEditAdapter;
import com.mintthursday.recipe.creation.stepsadapter.RecyclerRowMoveCallback;
import com.mintthursday.recipe.creation.stepsadapter.StepsAdapter;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class NewRecipeFragment extends Fragment implements SelectCategoryFragment.NoticeDialogListener {

    public static final String ARG_EDIT_RECIPE = "ARG_EDIT_RECIPE";
    Set<String> chooseData = new HashSet<>();
    private TextView nameRecipe;
    private TextView descriptionRecipe;
    private TextView qtyPortionRecipe;
    private TextView timeRecipe;
    private IngredientEditAdapter ingredientAdapter;
    private StepsAdapter stepsAdapter;
    private TextView editCategories;
    private long idRecipe = 0;

    public NewRecipeFragment() {
    }

    public static NewRecipeFragment newInstance(Recipe recipe) {
        NewRecipeFragment fragment = new NewRecipeFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(ARG_EDIT_RECIPE, recipe);
        fragment.setArguments(bundle);
        return fragment;
    }

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_new_recipe, container, false);
        Toolbar myToolbar = view.findViewById(R.id.myToolbar);
        ((AppCompatActivity) getActivity()).setSupportActionBar(myToolbar);
        ImageView closeWindow = view.findViewById(R.id.close);


        editCategories = view.findViewById(R.id.categories);
        View butCattegory = view.findViewById(R.id.outlinedInputCategory);
        View butIngredient = view.findViewById(R.id.outlinedInputIngredient);
        View butAddStep = view.findViewById(R.id.btnAddStep);
        View butSaveRecipe = view.findViewById(R.id.btnSaveRecipe);

        nameRecipe = view.findViewById(R.id.textInputName);
        descriptionRecipe = view.findViewById(R.id.textInputDescription);
        qtyPortionRecipe = view.findViewById(R.id.textInputQtyPortions);
        timeRecipe = view.findViewById(R.id.textInputTime);

        initRecyclerViewIngredients(view);
        initRecyclerViewSteps(view);
        loadSteps();
        if (getArguments() != null && getArguments().getParcelable(ARG_EDIT_RECIPE) != null) {
            Recipe recipe = getArguments().getParcelable(ARG_EDIT_RECIPE);
            initEditRecipe(recipe);
        }


        butCattegory.setOnClickListener(v -> {
            DialogFragment newFragment = SelectCategoryFragment.newInstance(chooseData);
            newFragment.show(getChildFragmentManager(), "abc");
        });
        butCattegory.setOnFocusChangeListener((view1, hasFocus) -> {
            if (view1.isInTouchMode() && hasFocus) {
                view1.performClick();  // picks up first tap
            }
        });
        butIngredient.setOnClickListener(v -> App.getInstance().getRouter().navigateTo(MainActivity.Screens.INSTANCE.showNewIngredient()));
        closeWindow.setOnClickListener(v -> requireActivity().onBackPressed());
        butAddStep.setOnClickListener(v -> addStep());
        butSaveRecipe.setOnClickListener(v -> {
            Recipe recipe = buildRecipe();
            saveRecipe(recipe);
            requireActivity().onBackPressed();
        });
        getParentFragmentManager().setFragmentResultListener(REQ_INGREDIENT, this, new FragmentResultListener() {
            @Override
            public void onFragmentResult(@NonNull String key, @NonNull Bundle bundle) {
                Ingredient result = bundle.getParcelable(BUN_INGREDIENT);
                if (result != null) {
                    Ingredient ingredient = new Ingredient(result.getName(), result.getQuantity(), result.getUnit());
                    int itemPosition = bundle.getInt(BUN_ITEM_POSITION);
                    if (itemPosition == -1) {
                        ingredientAdapter.addItem(ingredient);
                    } else {
                        ingredientAdapter.updateItem(ingredient, itemPosition);
                    }
                }
            }
        });
        return view;
    }

    private void initEditRecipe(Recipe recipe) {
        //todo не сохранеются шаги рцепта если открыть редактирование, добавить шшаг. нажать добавить ингредиент и вернуться
        // также с другими полямино тебе везет
        if (recipe != null) {
            idRecipe = recipe.getId();
            nameRecipe.setText(recipe.getName());
            descriptionRecipe.setText(recipe.getDescription());
            qtyPortionRecipe.setText(String.valueOf(recipe.getCountPortion()));
            timeRecipe.setText(String.valueOf(recipe.getTime()));
            editCategories.setText(recipe.getCategory());
            ingredientAdapter.setItems(recipe.getIngredients());
            List<String> list = recipe.getSteps();
            List<Step> steps = new ArrayList<>();
            for (int i = 0; i < list.size(); i++) {
                steps.add(new Step(list.get(i)));
            }
            stepsAdapter.setItems(steps);
        }

    }

    private void initRecyclerViewIngredients(View view) {
        RecyclerView ingrRecyclerView = view.findViewById(R.id.ingrRecyclerView);
        ingrRecyclerView.setLayoutManager(new LinearLayoutManager(this.getContext()));
        ingredientAdapter = new IngredientEditAdapter();
        ingrRecyclerView.setAdapter(ingredientAdapter);
        IngredientEditAdapter.OnIngredientClickListener l = ((ingredient, itemPosition) -> App.getInstance().getRouter().navigateTo(MainActivity.Screens.INSTANCE.editIngredient(ingredient, itemPosition)));
        ingredientAdapter.setOnItemClickListener(l);
    }

    private void initRecyclerViewSteps(View view) {
        RecyclerView stepsRecyclerView = view.findViewById(R.id.stepsRecyclerView);
        stepsRecyclerView.setLayoutManager(new LinearLayoutManager(this.getContext()));
        stepsAdapter = new StepsAdapter();
        ItemTouchHelper.Callback callback = new RecyclerRowMoveCallback(stepsAdapter);
        ItemTouchHelper touchHelper = new ItemTouchHelper(callback);
        touchHelper.attachToRecyclerView(stepsRecyclerView);
        stepsRecyclerView.setAdapter(stepsAdapter);
    }

    private void loadSteps() {
        List<Step> steps = getSteps();
        stepsAdapter.setItems(steps);
    }

    private List<Step> getSteps() {
        List<Step> list = new ArrayList<>();
        for (int i = 0; i < 3; i++) {
            list.add(new Step(""));
        }
        return list;
    }

    private void addStep() {
        Step step = getOneStep();
        stepsAdapter.addItem(step);
    }

    private Step getOneStep() {
        return new Step("");
    }

    @Override
    public void onDialogPositiveClick(DialogFragment dialog, Set<String> chooseData) {
        this.chooseData = chooseData;
        editCategories.setVisibility(View.VISIBLE);
        editCategories.setText(toStringSetWithJoin(chooseData));
    }

    @Override
    public void onDialogNegativeClick(DialogFragment dialog) {
    }

    private Recipe buildRecipe() {
        String name = String.valueOf(nameRecipe.getText());
        String description = String.valueOf(descriptionRecipe.getText());
        int countPortion = 0;
        int time = 0;
        try {
            countPortion = Integer.parseInt(qtyPortionRecipe.getText().toString());
            time = Integer.parseInt(timeRecipe.getText().toString());
        } catch (NumberFormatException e) {
            e.printStackTrace();
        }
        String category = String.valueOf(editCategories.getText());
        List<Ingredient> ingredients = ingredientAdapter.getItems();
        List<Step> steps = stepsAdapter.getItems();
        List<String> steps2 = new ArrayList<>();
        for (int i = 0; i < steps.size(); i++) {
            steps2.add(steps.get(i).getStepInstruction());
        }
        return new Recipe(idRecipe, name, description, countPortion, time, category, ingredients, steps2);

    }

    private void saveRecipe(Recipe recipe) {
        AppDatabase db = App.getInstance().getDatabase();
        RecipeDao recipeDao = db.recipeDao();
        recipeDao.insert(recipe);
    }

    public String toStringSetWithJoin(Set<String> set) {
        StringBuilder result = new StringBuilder();
        String divider = "";
        for (String s : set) {
            result.append(divider).append(s);
            if (divider.isEmpty()) {
                divider = ", ";
            }
        }
        return result.toString();
    }
}

File: app/src/main/java/com/mintthursday/recipe/creation/NewRecipeFragment.kt
package com.mintthursday.recipe.creation

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.ViewGroup
import androidx.core.os.bundleOf
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import androidx.navigation.findNavController
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.mintthursday.App
import com.mintthursday.R
import com.mintthursday.databinding.FragmentNewRecipeBinding
import com.mintthursday.models.Ingredient
import com.mintthursday.models.Recipe
import com.mintthursday.models.Step
import com.mintthursday.recipe.creation.SelectCategoryFragment.NoticeDialogListener
import com.mintthursday.recipe.creation.ingredientcreation.IngredientEditAdapter
import com.mintthursday.recipe.creation.ingredientcreation.IngredientEditAdapter.OnIngredientClickListener
import com.mintthursday.recipe.creation.ingredientcreation.IngredientFragment
import com.mintthursday.recipe.creation.stepsadapter.RecyclerRowMoveCallback
import com.mintthursday.recipe.creation.stepsadapter.StepsAdapter

class NewRecipeFragment : Fragment(), NoticeDialogListener {

    private val router = App.instance.router

    private var _binding: FragmentNewRecipeBinding? = null
    private val binding get() = _binding!!

    private val ingredientEditAdapter: IngredientEditAdapter = IngredientEditAdapter()
    private val stepsAdapter: StepsAdapter = StepsAdapter()
    private var idRecipe: Long = 0
    private var chooseCategories: Collection<String> = emptySet()

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelableArrayList(INGREDIENT_LIST, ArrayList(ingredientEditAdapter.items))
        outState.putParcelableArrayList(STEP_LIST, ArrayList(stepsAdapter.items))
        outState.putStringArrayList(CATEGORY_LIST, ArrayList(chooseCategories))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (savedInstanceState == null) {
            val recipe = arguments?.getParcelable<Recipe>(ARG_EDIT_RECIPE)
            if (recipe != null) {
                ingredientEditAdapter.setItems(recipe.ingredients)
                stepsAdapter.setItems(recipe.steps)
                chooseCategories = recipe.category.splitToSet()
            } else {
                stepsAdapter.setItems(startSteps())
            }
        } else {
            savedInstanceState.getParcelableArrayList<Ingredient>(INGREDIENT_LIST)?.let {
                ingredientEditAdapter.setItems(it)
            }
            savedInstanceState.getParcelableArrayList<Step>(STEP_LIST)?.let {
                stepsAdapter.setItems(it)
            }
            savedInstanceState.getStringArrayList(CATEGORY_LIST)?.let {
                chooseCategories = it.toSet()
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {

        _binding = FragmentNewRecipeBinding.inflate(inflater, container, false)

        binding.textInputName.addTextChangedListener(recipeTextWatcher)
        binding.textInputDescription.addTextChangedListener(recipeTextWatcher)
        binding.textInputQtyPortions.addTextChangedListener(recipeTextWatcher)
        binding.textInputTime.addTextChangedListener(recipeTextWatcher)
        binding.categories.addTextChangedListener(recipeTextWatcher)

        initRecyclerViewIngredients(binding)
        initRecyclerViewSteps(binding)

        if (arguments != null && requireArguments().getParcelable<Recipe>(ARG_EDIT_RECIPE) != null) {
            initEditRecipe(requireArguments().getParcelable(ARG_EDIT_RECIPE)!!)
        }

        setCategories(chooseCategories)

        binding.filledInputCategory.onFocusChangeListener = OnFocusChangeListener { v: View, hasFocus: Boolean ->
            if (v.isInTouchMode && hasFocus) {
                v.performClick() // picks up first tap
            }
        }

        binding.filledInputCategory.setOnClickListener {
            val newFragment: DialogFragment = SelectCategoryFragment.newInstance(chooseCategories)
            newFragment.show(childFragmentManager, DIALOG_CATEGORY)
        }

        binding.filledInputIngredient.onFocusChangeListener = OnFocusChangeListener { v: View, hasFocus: Boolean ->
            if (v.isInTouchMode && hasFocus) {
                v.performClick() // picks up first tap
            }
        }
        binding.filledInputIngredient.setOnClickListener {
            binding.root.findNavController().navigate(R.id.action_newRecipeFragment_to_ingredientFragment)
        }


        binding.close.setOnClickListener { router.exit() }//todo navigation
        binding.btnAddStep.setOnClickListener { addStep() }
        binding.btnSaveRecipe.setOnClickListener {
            saveRecipe(buildRecipe())
//            router.exit()
            binding.root.findNavController().navigateUp()
        }

        parentFragmentManager.setFragmentResultListener(IngredientFragment.REQ_INGREDIENT, this, { key, bundle ->
            val result: Ingredient? = bundle.getParcelable(IngredientFragment.BUN_INGREDIENT)
            if (result != null) {
                val ingredient = Ingredient(result.name, result.quantity, result.unit)
                val itemPosition = bundle.getInt(IngredientFragment.BUN_ITEM_POSITION)
                if (itemPosition == -1) {
                    ingredientEditAdapter.addItem(ingredient)
                } else {
                    ingredientEditAdapter.updateItem(ingredient, itemPosition)
                }
                changeVisibilityMinAmountOfIngredient()
            }
        })
        return binding.root
    }

    private fun changeVisibilityMinAmountOfIngredient() {
        if (ingredientEditAdapter.itemCount >= 3) {
            binding.ingredientmin.visibility = View.GONE
        } else {
            binding.ingredientmin.visibility = View.VISIBLE
        }
    }

    private fun initRecyclerViewIngredients(binding: FragmentNewRecipeBinding) {
        ingredientEditAdapter.setOnItemClickListener(object : OnIngredientClickListener {
            override fun onItemClick(ingredient: Ingredient, itemPosition: Int) {
//                router.navigateTo(Screens.editIngredient(ingredient, itemPosition))
                binding.root.findNavController().navigate(R.id.action_newRecipeFragment_to_ingredientFragment, bundleOf(IngredientFragment.ARG_INGREDIENT to ingredient, IngredientFragment.ARG_POSITION to itemPosition))
            }
        })
        binding.ingrRecyclerView.layoutManager = LinearLayoutManager(this.context)
        binding.ingrRecyclerView.adapter = ingredientEditAdapter
    }

    private fun initEditRecipe(recipe: Recipe) {
        idRecipe = recipe.id
        binding.textInputName.setText(recipe.name)
        binding.textInputDescription.setText(recipe.description)
        binding.textInputQtyPortions.setText(recipe.countPortion.toString())
        binding.textInputTime.setText(recipe.time.toString())
    }

    private fun String.splitToSet(delimiter: String = ", "): Set<String> {
        return this.split(delimiter).toSet()
    }

    private fun initRecyclerViewSteps(binding: FragmentNewRecipeBinding) {
        val stepsRecyclerView: RecyclerView = binding.stepsRecyclerView
        stepsRecyclerView.layoutManager = LinearLayoutManager(this.context)
        val callback: ItemTouchHelper.Callback = RecyclerRowMoveCallback(stepsAdapter)
        val touchHelper = ItemTouchHelper(callback)
        touchHelper.attachToRecyclerView(stepsRecyclerView)
        stepsRecyclerView.adapter = stepsAdapter
    }

    private fun startSteps(): List<Step> {
        return List(3, { Step("") })
    }

    private fun addStep() {
        stepsAdapter.addItem(Step(""))
    }

    override fun onDialogPositiveClick(dialog: DialogFragment, chooseData: Set<String>) {
        setCategories(chooseData)
    }

    private fun setCategories(categories: Collection<String>) {
        this.chooseCategories = categories
        binding.categories.visibility = if (categories.isEmpty()) View.GONE else View.VISIBLE
        binding.categories.text = categories.joinToString(", ")
    }

    override fun onDialogNegativeClick(dialog: DialogFragment) {}

    private fun buildRecipe(): Recipe {
        val name = binding.textInputName.text.toString()
        val description = binding.textInputDescription.text.toString()
        var countPortion = 0
        var time = 0
        try {
            countPortion = binding.textInputQtyPortions.text.toString().toInt()
            time = binding.textInputTime.text.toString().toInt()
        } catch (e: NumberFormatException) {
            e.printStackTrace()
        }
        val category = binding.categories.text.toString()
        val ingredients: List<Ingredient> = ingredientEditAdapter.items
        val steps: List<Step> = stepsAdapter.items
        val inputLink = binding.filledInputLink.text.toString()
        return Recipe(idRecipe, name, description, countPortion, time, category, ingredients, steps, inputLink)
    }

    private fun saveRecipe(recipe: Recipe) {
        val db = App.instance.database
        db.recipeDao().insertRecipe(recipe)
    }

    private val recipeTextWatcher: TextWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        }

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {

            val nameInput = binding.textInputName.text.toString().trim()
            val descriptionInput = binding.textInputDescription.text.toString().trim()
            val quantityInput = binding.textInputQtyPortions.text.toString().trim()
            val timeInput = binding.textInputTime.text.toString().trim()
            val categoryInput = binding.categories.text.toString().trim()

            binding.btnSaveRecipe.isEnabled = nameInput.isNotEmpty() &&
                    descriptionInput.isNotEmpty() &&
                    quantityInput.isNotEmpty() &&
                    timeInput.isNotEmpty() &&
                    categoryInput.isNotEmpty()
        }

        override fun afterTextChanged(s: Editable?) {}

    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null //todo зачем мы это сделали?
    }

    companion object {

        const val ARG_EDIT_RECIPE = "ARG_EDIT_RECIPE"
        const val INGREDIENT_LIST = "INGREDIENT_LIST"
        const val STEP_LIST = "STEP_LIST"
        const val CATEGORY_LIST = "CATEGORY_LIST"
        const val DIALOG_CATEGORY = "DIALOG_CATEGORY"


//        fun newInstance(recipe: Recipe): NewRecipeFragment {
//            val fragment = NewRecipeFragment()
//            val bundle = Bundle()
//            bundle.putParcelable(ARG_EDIT_RECIPE, recipe)
//            fragment.arguments = bundle
//            return fragment
//        }
    }
}